### PR TITLE
JDBC_PING as default discovery protocol

### LINKED ISSUE #29399: JDBC_PING2 as default discovery protocol
### Description

The default discovery protocol provided by Infinispan uses IP multicast (`MPING`) to find its peers in the network. Although it is perfectly fine for an environment with IP multicast available, `JDBC_PING2` seems a better fit as a default discovery protocol for Keycloak. 

The `JDBC_PING2` protocol uses a database to keep track of the peer's IP addresses (and ports) and, we can use the Keycloak database as a registry.

The following work is required:

* Create a custom stack, replacing `MPING` with `JDBC_PING2`
* Infinispan 15.0.3.Final allows injecting a data source into `JDBC_PING2` for Infinispan Server (see https://issues.redhat.com/browse/ISPN-15961). The same code can be used by Keycloak to inject the Quarkus data source into JGroups.
* (UX improvement) Somehow detect the database dialect to configure the `JDBC_PING2` queries (select, delete, insert). The goal is to achieve zero-config for users.
* Set values for remaining options based on user feedback and/or issues found.

`JDBC_PING2` documentation is available here: http://jgroups.org/manual5/index.html#JDBC_PING2

### Discussion

https://github.com/keycloak/keycloak/discussions/29205

### Issues

_No response_

### Motivation

To simplify Keycloak usage and configuration.

### PR DESCRIPTION
**Pull request by ryanemerson**
Closes #29399

This can't be merged until Keycloak 26.1, but I'm creating a draft in order to receive feedback.

This PR makes it so that when a Keycloak cluster is created with an Embedded Infinispan cluster and the `--cache-stack` arg is not explicitly provided, the default discovery protocol is configured to be `JDBC_PING2` using the Keycloak DB.  When deploying to K8s environments, it's still preferable to utilise the `DNS_PING` discovery protocol, but as the operator already utilises `--cache-stack=kubernetes` this behaviour should remain unchanged.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -588,6 +588,16 @@ jobs:
           echo "Tests: $TESTS"
           ./mvnw test ${{ env.SUREFIRE_RETRY }} -Pauth-server-quarkus -Pdb-${{ matrix.db }} "-Dwebdriver.chrome.driver=$CHROMEWEBDRIVER/chromedriver" -Dtest=$TESTS -pl testsuite/integration-arquillian/tests/base 2>&1 | misc/log/trimmer.sh
 
+      - name: Run cluster JDBC_PING2 smoke test
+        run: |
+          ./mvnw test ${{ env.SUREFIRE_RETRY }} \
+            -Pauth-server-cluster-quarkus \
+            -Pdb-${{ matrix.db }} \
+            -Dtest=RealmInvalidationClusterTest \
```
Comment (ryanemerson) [inline .github/workflows/ci.yml:596]: I am re-using an existing cluster test here for simplicity, but I can add a dedicated smoke cluster test which just ensures that a cluster of size `x` is created as expected. The important thing is that we ensure a cluster is able to form with each of Keycloak's supported DBs.

More thorough testing of the JDBC_PING2 protocol and how it handles is cluster members leaving gracefully/abrubtly existing in the Infinispan repo [JdbcPing2IT](https://github.com/infinispan/infinispan/blob/main/server/tests/src/test/java/org/infinispan/server/persistence/JdbcPing2IT.java) [[JGroupsJdbcPing2IT.java](https://github.com/infinispan/infinispan/blob/main/server/tests/src/test/java/org/infinispan/server/persistence/JGroupsJdbcPing2IT.java)](https://github.com/infinispan/infinispan/blob/main/server/tests/src/test/java/org/infinispan/server/persistence/[JGroupsJdbcPing2IT.java](https://github.com/infinispan/infinispan/blob/main/server/tests/src/test/java/org/infinispan/server/persistence/JGroupsJdbcPing2IT.java))

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -588,6 +588,16 @@ jobs:
           echo "Tests: $TESTS"
           ./mvnw test ${{ env.SUREFIRE_RETRY }} -Pauth-server-quarkus -Pdb-${{ matrix.db }} "-Dwebdriver.chrome.driver=$CHROMEWEBDRIVER/chromedriver" -Dtest=$TESTS -pl testsuite/integration-arquillian/tests/base 2>&1 | misc/log/trimmer.sh
 
+      - name: Run cluster JDBC_PING2 smoke test
```
Comment (ryanemerson) [inline .github/workflows/ci.yml:591]: I have added this to the existing `store-integration-tests` job for simplicity, but maybe it's better to create a dedicated job as this is not really related to store functionality?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java [hunk 3]
```diff
@@ -99,9 +100,12 @@ public void init() {
         if (componentFactoryPF != null) {
             componentFactoryPF.postInit(this);
         }
+        // Init JpaConnectionProvider first to ensure that DB resources are available to all other providers
+        ProviderFactory<JpaConnectionProvider> jpaUpdaterProvider = getProviderFactory(JpaConnectionProvider.class);
+        jpaUpdaterProvider.postInit(this);
```
Comment (ryanemerson) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java:105]: @keycloak/cloud-native I would like advice here.

With JDBC_PING2, the Infinispan cache manager now has a hard dependency on the JDBCPING DB table and we need to ensure that all DB creations/updates occur before the cache manager is created. The current workflow seems to be that the updates are initiated by `QuarkusJpaUpdaterProvider#update` via `QuarkusJpaConnectionProviderFactory#postInit`, hence my current code which ensures the `postInit` method on this ProviderFactory is called first. This solution works, but seems a bit hacky and I wondered if there was a better approach to solving this issue?

After discussing with @mhajas, I have tried to use Quarkus BuildItem's with the `CacheBuildSteps` to force a before/after relationship with Liquibase and Jpa, but this does not seem to work as ultimately a call to `postInit` on the factory is what initiates the migration.

Any thoughts/suggestions much appreciated :slightly_smiling_face:

**Comment by ahus1** [comment]
About the "order of things" in relation to Infinispan and the database: In another train of thought, we want to store credentials and certificates needed for the JGroups communication in the database. Those would require the database to be initialized first as well.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java [hunk 4]
```diff
@@ -361,6 +377,28 @@ private static void configureTransportStack(ConfigurationBuilderHolder builder)
...
+                  "delete_single_sql", String.format("DELETE from %s WHERE address=?", tableName),
+                  "insert_single_sql", String.format("INSERT INTO %s values (?, ?, ?, ?, ?)", tableName),
+                  "select_all_pingdata_sql", String.format("SELECT address, name, ip, coord FROM %s WHERE cluster_name=?", tableName),
+                  "remove_all_data_on_view_change", "true",
+                  "stack.combine", "REPLACE",
+                  "stack.position", "PING"
+            );
+            var stackName = "keycloak-jdbc-ping";
+            var stack = List.of(new ProtocolConfiguration(JDBC_PING2.class.getSimpleName(), attributes));
+            builder.addJGroupsStack(new EmbeddedJGroupsChannelConfigurator(stackName, stack, null), "udp");
```
Comment (pruivo) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: Why not add the stack to the cache-ispn.xml? UDP is most likely blocked in some cloud environments, shouldn't be based on TCP?

Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: > Why not add the stack to the cache-ispn.xml?

Good point. Originally I thought that we were going to have to provide per vendor SQL, but now that's not the case I can define this declaratively.



> UDP is most likely blocked in some cloud environments, shouldn't be based on TCP?

The Keycloak default is `UDP`, which is why I have used that here. The Operator explicitly configures the transport stick as it relies on DNS_PING, so JGROUPSPING will never be used in that environment. Maybe there's an argument for changing the default from UDP to TCP, but I think that needs to be a separate issue.

Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: Or add 2 stacks, one for UDP and TCP. 
Are we changing the default to use JDBC_PING or keeping plain UDP?

Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: The aim of the JDBC_PING was to have something that works out-of-the-box in all places. As we use TCP in Kubernetes, and have tested it a lot, I'd +1 to move to TCP as well to have one thing less users need to worry about.

Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: I've created https://github.com/keycloak/keycloak/issues/33651 to change the default to TCP.

Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: > > Why not add the stack to the cache-ispn.xml?
> 
> Good point. Originally I thought that we were going to have to provide per vendor SQL, but now that's not the case I can define this declaratively.

A downside to defining the stack in `cache-ispn.xml` is that it's then not possible to use JDBC_PING2 with a custom `--cache-config-file`. For example they may want to use the default stacks, but provide explicit configuration of the caches. The current approach allows them to do this, but they would have to also provide `cache-stack` if they wanted to deviate from JDBC_PING2.

**Review by ahus1 — CHANGES_REQUESTED** [review]
I think the approach will have a problem with DB schema changes, but this is IMHO fixable. See below.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java (hunk 4 above)
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:397]: Fair enough. I'm not going to enter into the "what if" spiral :rofl: 
It needs documentation that says that the stack is automatically available.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-26.0.0.xml [hunk 5]
```diff
@@ -113,4 +113,21 @@
...
+            <column name="cluster_name" type="VARCHAR(200)">
+                <constraints nullable="false" />
+            </column>
+            <column name="ip" type="VARCHAR(200)">
+                <constraints nullable="false" />
+            </column>
+            <column name="coord" type="BOOLEAN"/>
+        </createTable>
+        <addPrimaryKey columnNames="address" constraintName="CONSTRAINT_JGROUPSPING" tableName="JGROUPSPING"/>
+    </changeSet>
```
Comment (ahus1) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-26.0.0.xml:131]: With KC26.0 being released, this needs to go to a new file 26.1.0

**Comment by ahus1** [comment]
With 26.1 (or any 26.x), we can't make this the default discovery protocol, but just "another" discovery protocol. Only for 27.0 we can make this then the default. When deciding between preview and supported, I'd say we can aim for supported, can we?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java [hunk 6]
```diff
@@ -355,11 +375,34 @@ private static int getStartTimeout() {
...
     }
 
-    private static void configureTransportStack(ConfigurationBuilderHolder builder) {
+    private static void configureTransportStack(ConfigurationBuilderHolder builder, KeycloakSession keycloakSession) {
         String transportStack = Configuration.getRawValue("kc.cache-stack");
 
         var transportConfig = builder.getGlobalConfigurationBuilder().transport();
         if (transportStack != null && !transportStack.isBlank()) {
+            if ("jdbc-ping".equals(transportStack)) {
+                EntityManager em = keycloakSession.getProvider(JpaConnectionProvider.class).getEntityManager();
```
Comment (mhajas) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:384]: @ryanemerson Would this line be enough to make sure the tables are created at the time jdbc ping is initialized? Maybe we don't need anything to change in `quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java`

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java (hunk 3 above)
Comment (stianst) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java:105]: Think we need some priorities or dependencies between factories, so that we initialize things in the correct order. This is a bit to hacky if you ask me ;)

**Comment by stianst** [comment]
How does JDBC_PING affect startup time? The default stack today results in a couple seconds extra startup time, hoping that JDBC_PING doesn't have this problem.

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
> How does JDBC_PING affect startup time? The default stack today results in a couple seconds extra startup time, hoping that JDBC_PING doesn't have this problem.

@stianst The only delay is querying execution duration. If the result set is empty, KC proceeds with the startup. Otherwise, it sends a join request to the coordinator. In any case, the RPC and the Database query combined should take a couple of milliseconds or less.

If the database is offline, then you have to take into consideration the SQL connection timeout.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java (hunk 6 above)
Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:384]: Unfortunately not. Removing my hack means that the `EntityManagerFactory` in the can be null and we get the following fatal exception:

```
2024-10-09 11:55:55,629 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
2024-10-09 11:55:55,630 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Cannot invoke "jakarta.persistence.EntityManagerFactory.createEntityManager(jakarta.persistence.SynchronizationType)" because "emf" is null
```

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java (hunk 3 above)
Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java]: I've added an optional `postInitPriority` method to the `ProviderFactory` which is used to sort the ProviderFactory implementations before executing `postInit`. The upshot is we can also remove the explicit `ComponentFactoryProviderFactory::postInit` call by returning `Integer.MAX_VALUE`. Maybe this is as equally as hacky as my initial attempt though :slightly_smiling_face:

**Comment by ryanemerson** [comment]
After discussing this issue further, we have decided to enable this by default in KC 26.1 after all. As we're only modifying the discovery protocol of the default "udp" stack, there shouldn't be any changes required for users upgrading from 26.0 to 26.1 as all configuration is handled internally. In the event that the old behaviour is truly desired, users can re-enable it by specifying `--cache-stack=udp`.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java (hunk 3 above)
Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java:105]: Actually that broke a ton of things :sweat_smile: , reverted, I'll wait for input from the Pros.

Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusKeycloakSessionFactory.java:105]: I agree @vmuzikar. I experimented a bit more locally to see why adding a priority to the postInit order did not work and AFAICT it seems that we're relying on the JVM ordering of the HashMap entries atm and it just happens to work, which of course may change in subsequent updates or with different JVMs. So I think a more robust mechanism is definitely required.

**Comment by ahus1** [comment]
@ryanemerson and me had a good code pairing session today to explore how factory providers can declare dependencies on other providers. We got a first prototype working that worked for starting up Keycloak locally. 

Ryan will have a look if all tests pass and what further changes to the non-Keycloak session factory are necessary.

File: server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java [hunk 7]
```diff
@@ -67,4 +69,13 @@ default int order() {
...
         return Collections.emptyList();
     }
+
+    /**
+     * Optional method used to declare that a ProviderFactory has a dependency on one or more Providers. If a Provider
+     * is declared here, it is guaranteed that the dependencies {@link #postInit} method will be executed
+     * before this ProviderFactory's {@link #postInit}.
+     */
+    default Set<Class<? extends Provider>> dependsOn() {
+        return Set.of();
```
Comment (stianst) [inline server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java:79]: Should this not use `Collections.emptySet()`?

File: model/jpa/src/main/resources/META-INF/jpa-changelog-26.1.0.xml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  ~ * Unless required by applicable law or agreed to in writing, software
+  ~ * distributed under the License is distributed on an "AS IS" BASIS,
+  ~ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ * See the License for the specific language governing permissions and
+  ~ * limitations under the License.
+  -->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <changeSet author="keycloak" id="29399-jdbc-ping-default">
+        <createTable tableName="JGROUPSPING">
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-26.1.0.xml:21]: Nitpick for sure, but maybe it should be `JGROUPS_PING`?

File: model/jpa/src/main/resources/META-INF/jpa-changelog-26.1.0.xml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+            <column name="name" type="VARCHAR(200)" />
+            <column name="cluster_name" type="VARCHAR(200)">
+                <constraints nullable="false" />
+            </column>
+            <column name="ip" type="VARCHAR(200)">
+                <constraints nullable="false" />
+            </column>
+            <column name="coord" type="BOOLEAN"/>
+        </createTable>
+        <addPrimaryKey columnNames="address" constraintName="CONSTRAINT_JGROUPSPING" tableName="JGROUPSPING"/>
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-26.1.0.xml:34]: Same nitpick as above `CONSTRAINT_JGROUPS_PING`?

**Review by stianst — COMMENTED** [review]
Is there some way we could unit tests the provider initialization? The logic is for sure getting a bit more complex now.

Also, if I remember correctly back in the day we only had `init`, then `postInit` was kinda added as a work-around to this same problem. Basically the idea was that in `init` you'd initialize the provider, while in `postInit` you could do additional init that depended on other providers. Honestly, having ability to set dependencies is a much better way, so maybe this would allow us to deprecate `postInit`, maybe make it a default empty method on `ProviderFactory`? Not something that needs to be part of this PR for sure, but maybe we could at least create a follow-up issue for that?

**Comment by stianst** [comment]
If I read the code correctly the dependencies on a factory is taken into account to decide on the order of calling `postInit`, but maybe this should also apply to `init` as well?

**Comment by stianst** [comment]
One thing this could open up to as well is ability to concurrently initialise provider factories for faster startups. That would require some more careful consideration on dependencies between things though, and for sure a follow-up, but perhaps we should have a follow-up to review providers to make sure all dependencies are listed?

File: services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java [hunk 10]
```diff
@@ -117,25 +118,56 @@ public void init() {
...
 
         AdminPermissions.registerListener(this);
     }
 
+    protected void initProviderFactories() {
+        // Component factory must be initialized first, so that postInit in other factories can use component factories
+        updateComponentFactoryProviderFactory();
+        if (componentFactoryPF != null) {
+            componentFactoryPF.postInit(this);
+        }
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java:134]: Never fully understood what component factories where, and why they needed to be initialised early. Seems that really this could/should be refactored to use dependencies as well?

File: server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java (hunk 7 above)
Comment (ryanemerson) [reply server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java:79]: I don't think we want the returned set to be mutable. `Set.of()`  just calls `ImmutableCollections.EMPTY_SET`.

File: services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java (hunk 10 above)
Comment (ryanemerson) [reply services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java:134]: I can experiment with removing this and see what happens without calling this first in all cases. Worst case, assuming this breaks a lot of things, a simple solution would be to make the default `ProviderFactory#dependsOn` method return a singleton set `ComponentFactoryProviderFactory`. Subsequent implementations could then add to this set, whereas the `ComponentFactoryProviderFactory#dependsOn` implementation would return an empty set as it obviously can't depend on itself.

File: server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java (hunk 7 above)
Comment (stianst) [reply server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java:79]: emptySet is immutable as well, but yeah if Set.of() returns the same then it doesn't really matter, an extra indirection though compared to just explicitly saying empty set (which is also a little bit more readable).

Nitpick though ;)

File: services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java (hunk 10 above)
Comment (stianst) [reply services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java:134]: Actually not sure AmphibianProviderFactory or ComponentFactoryProvider is even used anymore. It was something that was introduced for the new store, but now that's gone I don't think there's anything using it anymore.

File: services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java [hunk 11]
```diff
@@ -150,16 +184,17 @@ protected Map<Class<? extends Provider>, Map<String, ProviderFactory>> getFactor
...
-            Map<String, ProviderFactory> current = copy.get(entry.getKey());
+            Class<? extends Provider> provider = entry.getKey();
+            Map<String, ProviderFactory> current = copy.get(provider);
             if (current == null) {
-                copy.put(entry.getKey(), entry.getValue());
+                copy.put(provider, entry.getValue());
             } else {
                 for (ProviderFactory f : entry.getValue().values()) {
-                    deployed.add(f);
+                    deployed.put(provider, entry.getValue());
```
Comment (mhajas) [inline services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java:197]: Seems like this line can be moved out of the for loop.

**Comment by ryanemerson** [comment]
> > If I read the code correctly the dependencies on a factory is taken into account to decide on the order of calling `postInit`, but maybe this should also apply to `init` as well?
> 
> Sure, I should be able to easily add this as part of this PR.

I was wrong on this one :slightly_smiling_face:. I tried to unify calls so that we would respect `dependsOn` in the initialization order, and then call `init` followed by `postInit` for each ProviderFactory. Unfortunately though many `postInit` implementations depend on `init` having already been called for other ProviderFactory implementations, so this effort is non-trivial and I think beyond the scope of this PR.

> Actually not sure AmphibianProviderFactory or ComponentFactoryProvider is even used anymore. It was something that was introduced for the new store, but now that's gone I don't think there's anything using it anymore.

I also [tried to remove these classes](https://github.com/keycloak/keycloak/compare/main...ryanemerson:keycloak:JDBC_PING2_remove_cf), however this causes issues with providers not found:

```
2024-10-16 14:41:42,161 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: java.lang.IllegalStateException: Executor with provider ID secure-session not found
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getExecutorProvider(ClientPoliciesUtil.java:161)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.isValidExecutor(ClientPoliciesUtil.java:349)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getValidatedGlobalClientProfilesRepresentation(ClientPoliciesUtil.java:213)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.readGlobalClientProfilesRepresentation(ClientPoliciesUtil.java:83)
	at org.keycloak.services.securityprofile.DefaultSecurityProfileProviderFactory.readConfiguration(DefaultSecurityProfileProviderFactory.java:105)
	at org.keycloak.services.securityprofile.DefaultSecurityProfileProviderFactory.create(DefaultSecurityProfileProviderFactory.java:52)
	at org.keycloak.services.securityprofile.DefaultSecurityProfileProviderFactory.create(DefaultSecurityProfileProviderFactory.java:43)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:178)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getGlobalClientPolicies(ClientPoliciesUtil.java:382)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getEnabledClientPolicies(ClientPoliciesUtil.java:392)
	at org.keycloak.services.clientpolicy.DefaultClientPolicyManager.doPolicyOperation(DefaultClientPolicyManager.java:64)
	at org.keycloak.services.clientpolicy.DefaultClientPolicyManager.triggerOnEvent(DefaultClientPolicyManager.java:56)
	at org.keycloak.protocol.oidc.grants.ResourceOwnerPasswordCredentialsGrantType.process(ResourceOwnerPasswordCredentialsGrantType.java:82)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:151)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint$quarkusrestinvoker$processGrantRequest_6408e15340992839b66447750c221d9aaa837bd7.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

I think this is beyond the scope of this PR (and my knowledge of Keycloak). I have added both items as potential refactoring items in the #33887 epic which may be of interest to the core team.

File: services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java (hunk 10 above)
Comment (ryanemerson) [reply services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java]: ^ is not possible due to `ComponentFactoryProviderFactory` being part of the `keycloak-server-spi-private` module and `ProviderFactory` being in `keycloak-server-spi`.

**Comment by ahus1** [comment]
@ryanemerson - I had a look at the current state today, and I think there are still two issues: 

* When using a multi-site setup, both sites will use the same database. As of today, IMHO they are both writing ISPN as the clustername to the DB, so they will interfere. For multi-site, we have `spi-connections-infinispan-quarkus-site-name` already which IMHO could or even should be used here. 

* When shutting down the dev environment with H2, the connection pool complains that the database has already been closed. So it seems to me we need to try other things than a manual shutdown. One of the options might be to have `DB_CLOSE_DELAY=0` (which is AFAIK the default), and then ensure hat the min pool size when using H2 is set to 1 instead of the default 0.

**Comment by pruivo** [comment]
@ahus1 site name and cluster name are 2 different things and it is better to use two different CLI options. For a stretched cluster, the cluster name is the same for all KC nodes while the site name is different and dependent on the node location. 

I strongly advise not using the site name as a cluster name; if we want to support a stretched cluster, it would require some kind of migration guide to tell users to stop using the site name and use the cluster name.

**Comment by ahus1** [comment]
> @ahus1 site name and cluster name are 2 different things and it is better to use two different CLI options. For a stretched cluster, the cluster name is the same for all KC nodes while the site name is different and dependent on the node location.
> 
> I strongly advise not using the site name as a cluster name; if we want to support a stretched cluster, it would require some kind of migration guide to tell users to stop using the site name and use the cluster name.

Thank you for raising this. Let's discuss this further in #34160 where we want to have a proper CLI option. 

In the meantime using the site name would serve IMHO as good-enough fallback for those doing multisite today in a non-Kubernetes setup to make it backwards compatible which is a requirement for including it in a minor release. 

Once the proper CLI option is there, we can add for example warnings when this fallback issue is used.

**Comment by ryanemerson** [comment]
> I strongly advise not using the site name as a cluster name; if we want to support a stretched cluster, it would require some kind of migration guide to tell users to stop using the site name and use the cluster name.

The current use of the site provider is nested within a `MultiSiteUtils#isMultiSiteEnabled`. I assume that support for stretch clusters will only be possible when `!MULTI_SITE` and/or some additional feature is enabled to indicate stretch clusters are required, so I think we should be able to avoid the conflict you describe.

**Comment by pruivo** [comment]
No feature is required for a stretched cluster. It only needs to configure the Infinispan transport: 
```java
transportConfig.siteId("az1")
```
If you use the site name as a cluster name, what would be the name of the CLI options to configure the site-id? 

The `spi-connections-infinispan-quarkus-site-name` is no longer required with `multi-site` and it was used for the event with KC24, to avoid processing the events twice.

**Comment by ryanemerson** [comment]
> No feature is required for a stretched cluster.

Great, that works in our favour. The use of site-name here is hidden behind MULTI_SITE, so if the user requires a stretched cluster they won't have the MULTI_SITE feature enabled.

> If you use the site name as a cluster name, what would be the name of the CLI options to configure the site-id?

We're using the site-id only as a suffix for the cluster name, so the cluster name and site name won't be the same.

> The `spi-connections-infinispan-quarkus-site-name` is no longer required with `multi-site` and it was used for the event with KC24, to avoid processing the events twice.

It's still a requirement according to the docs, so I think it's safe to depend on it being set for this edge-case.

**Comment by ahus1** [comment]
> The `spi-connections-infinispan-quarkus-site-name` is no longer required with `multi-site` and it was used for the event with KC24, to avoid processing the events twice.

Again, we can use the other issue to remove it. Still, it is good that we have it in the docs otherwise it would be a breaking change. :rofl:

**Comment by ryanemerson** [comment]
For others interested in this PR: after further discussion @pruivo explained that our concerns around JDBC_PING and multi-site were in fact moot, as the MULTI_SITE feature requires an external Infinispan cluster, not the embedded one which we're configuring JDBC_PING for. I have removed the relevant code.

**Review by ahus1 — APPROVED** [review]
Thank you for this change. LGTM, and all previous review comments have been addressed.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java [hunk 12]
```diff
@@ -357,12 +370,37 @@ private static int getStartTimeout() {
...
+                  "select_all_pingdata_sql", String.format("SELECT address, name, ip, coord FROM %s WHERE cluster_name=?", tableName),
+                  "remove_all_data_on_view_change", "true",
+                  "register_shutdown_hook", "false",
+                  "stack.combine", "REPLACE",
+                  "stack.position", "PING"
+            );
+            var stack = List.of(new ProtocolConfiguration(JDBC_PING2.class.getSimpleName(), attributes));
+            builder.addJGroupsStack(new EmbeddedJGroupsChannelConfigurator(jdbcStackName, stack, null), "udp");
+
+            Supplier<DataSource> dataSourceSupplier = Arc.container().select(AgroalDataSource.class)::get;
```
Comment (yanxch) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:401]: Afterthought:

Because I just discovered that our jdpc_ping configuration in the cache-config.xml does not use a connection pool.

It would have been nice to create a specific class that could be reused by folks who already have jdbc_ping activated in their cache config.

Comment (ryanemerson) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:401]: @yanxch Can you explain why you need a custom JDBC_PING configuration now that it is supported natively by Keycloak?

Comment (yanxch) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/CacheManagerFactory.java:401]: I didn't check the other configs we currently have in place, they may differ from the one Keycloak is providing.
But you'r right. Egde Case
